### PR TITLE
feat: add support for custom cache stores (`options.read/options.write`)

### PR DESCRIPTION
**Pull request by Mordred**
Hi, I want to change how cache data are stored in my project. In my case I want to use Redis server like this:

```js
const redis = require('redis');

// ...
// ... connect to client
// ...

const BUILD_CACHE_TIMEOUT = 24 * 3600; // 1 day

function redisKey(filename) {
    return `build:cache:${filename}`;
}

function readCache(filename, callback) {
      client.get(redisKey(filename), (err, result) => {
          if (err) {
              return callback(err);
          }

          if (!result) {
              return callback(new Error(`Key ${redisKey(filename)} not found`));
          }

          callback(null, result);
    });
}

function writeCache(filename, content, callback) {
      client.set(redisKey(filename), content, 'EX', BUILD_CACHE_TIMEOUT, callback);
}
```

Then in my webpack config:
```js
{
  loader: 'cache-loader',
  options: {
    readCache,
    writeCache,
  },
},
```

For that I needed `writeCache` and `readCache` hooks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,6 +49,8 @@ module.exports = {
 |:--:|:--:|:-----:|:----------|
 |**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored|
 |**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.|
+|**`writeCache`**|`{Function(cacheFilename, content, callback) -> {void}}`|`undefined`|Allows you to override default write cache data to file (e.g. Redis, memcached).|
```
Comment (michael-ciniawsky) [inline README.md]: Just `writer` (`options.writer`) ?

File: README.md [hunk 2]
```diff
@@ -49,6 +49,8 @@ module.exports = {
...
+|**`readCache`**|`{Function(cacheFilename, callback) -> {void}}`|`undefined`|Allows you to override default read cache data from file.|
```
Comment (michael-ciniawsky) [inline README.md]: Just `reader` (`options.reader`) ?

File: README.md [hunk 3]
```diff
@@ -49,6 +49,8 @@ module.exports = {
...
 
 <h2 align="center">Examples</h2>
```
Comment (michael-ciniawsky) [inline README.md]: Please add an example aswell

**Comment by Mordred** [comment]
@michael-ciniawsky `fs` is faster than database, but we are building application for production in the docker. Saving cached data to filesystem is not an option because mounting volumes during `docker build` is not possible.

I have changed option names and added example from my first post.

File: README.md [hunk 4]
```diff
@@ -95,6 +97,62 @@ module.exports = {
 }
 ```
 
+### `Write cache data to redis`
```
Comment (michael-ciniawsky) [inline README.md]: Could you please try to make this example a bit more generic? e.g `Database Intergation` something like that :)

**Comment by sokra** [comment]
cache-loader should no longer talk about filenames. Filenames should only occur in the default read/write/key implementation.

**Comment by Mordred** [comment]
@sokra refactored

File: README.md [hunk 5]
```diff
@@ -95,6 +98,82 @@ module.exports = {
 }
 ```
 
+### `Example of Database Integration`
```
Comment (michael-ciniawsky) [inline README.md]: `### Database Intergration`

File: src/index.js [hunk 6]
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
 const defaultCacheDirectory = path.resolve('.cache-loader');
 const ENV = process.env.NODE_ENV || 'development';
 
+const defaultOptions = {
```
Comment (michael-ciniawsky) [inline src/index.js]: `defaultOptions` => `defaults`

File: src/index.js [hunk 7]
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
 const defaultCacheDirectory = path.resolve('.cache-loader');
```
Comment (michael-ciniawsky) [inline src/index.js]:
```diff
- const defaultCacheDirectory = path.resolve('.cache-loader');
```

File: src/index.js [hunk 8]
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
...
 const ENV = process.env.NODE_ENV || 'development';
 
+const defaultOptions = {
+  cacheDirectory: defaultCacheDirectory,
```
Comment (michael-ciniawsky) [inline src/index.js]:
```
cacheDirectory: path.resolve('.cache-loader'),
```

File: src/index.js [hunk 9]
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
...
+  cacheIdentifier: `cache-loader:${pkgVersion} ${ENV}`,
+  read,
+  write,
+  makeKey,
+};
+
 function loader(...args) {
+  const loaderOptions = loaderUtils.getOptions(this) || {};
```
Comment (michael-ciniawsky) [inline src/index.js]:
```diff
- const loaderOptions = loaderUtils.getOptions(this) || {};
```

File: src/index.js [hunk 10]
```diff
@@ -9,7 +9,18 @@ const pkgVersion = require('../package.json').version;
...
+  const options = Object.assign({}, defaultOptions, loaderOptions);
```
Comment (michael-ciniawsky) [inline src/index.js]:
```js
const options = Object.assign({}, defaults, loaderUtils.getOptions(this))
```

File: src/index.js [hunk 11]
```diff
@@ -35,62 +46,33 @@ function loader(...args) {
...
   const callback = this.async();
-  const hash = digest(`${cacheIdentifier}\n${remainingRequest}`);
-  const cacheFile = path.join(cacheDirectory, `${hash}.json`);
+  const cacheKey = makeKeyFn(options, remainingRequest);
   data.remainingRequest = remainingRequest;
-  data.cacheIdentifier = cacheIdentifier;
-  data.cacheFile = cacheFile;
-  fs.readFile(cacheFile, 'utf-8', (readFileErr, content) => {
-    if (readFileErr) {
+  data.cacheKey = cacheKey;
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n data.cacheKey = cacheKey(options, remainingRequest) \n`

File: src/index.js [hunk 12]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
   return crypto.createHash('md5').update(str).digest('hex');
 }
 
+const directoriesExists = {};
```
Comment (michael-ciniawsky) [inline src/index.js]: `const directories = new Set();\n`

File: src/index.js [hunk 13]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
...
+function write(cacheKey, data, callback) {
+  const dirname = path.dirname(cacheKey);
+  const content = JSON.stringify(data);
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

File: src/index.js [hunk 14]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
...
+  if (directoriesExists[dirname]) {
```
Comment (michael-ciniawsky) [inline src/index.js]: `if (directories.has(dirname)) {`

File: src/index.js [hunk 15]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
...
+    // for performance skip creating directory
+    fs.writeFile(cacheKey, content, 'utf-8', callback);
+  } else {
+    mkdirp(dirname, (mkdirErr) => {
+      if (mkdirErr) {
+        callback(mkdirErr);
+        return;
+      }
+      directoriesExists[dirname] = true;
```
Comment (michael-ciniawsky) [inline src/index.js]: `\ndirectories.add(dirname)`

File: src/index.js [hunk 16]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
...
+    try {
+      const data = JSON.parse(content);
+      callback(null, data);
+    } catch (e) {
+      callback(e);
+    }
+  });
+}
+
+function makeKey(options, request) {
```
Comment (michael-ciniawsky) [inline src/index.js]: `function cacheKey(options, request) {`

File: src/index.js [hunk 17]
```diff
@@ -123,4 +105,45 @@ function digest(str) {
...
+  const { cacheIdentifier, cacheDirectory } = options;
+  const hash = digest(`${cacheIdentifier}\n${request}`);
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

File: src/index.js [hunk 18]
```diff
@@ -35,62 +46,33 @@ function loader(...args) {
...
 function pitch(remainingRequest, prevRequest, dataInput) {
   const loaderOptions = loaderUtils.getOptions(this) || {};
-  const defaultOptions = {
-    cacheDirectory: defaultCacheDirectory,
-    cacheIdentifier: `cache-loader:${pkgVersion} ${ENV}`,
-  };
   const options = Object.assign({}, defaultOptions, loaderOptions);
-  const { cacheIdentifier, cacheDirectory } = options;
+  const { read: readFn, makeKey: makeKeyFn } = options;
   const data = dataInput;
```
Comment (michael-ciniawsky) [inline src/index.js]: Please move this to L67

**Comment by Crazymax11** [comment]
Hello
We also tried to use cache loader to build application in docker and met two problems:
* Cache loader uses mtime to check if dependency changed, but git doesn't store files mtime, so after git clone we always have new mtime and no cache used.
* It seems that cache loader doesn't use content of resource to check if resources changed, cache loader just checks if dependencies changed.

It would be nice to know if module changed by checking hash content instead of mtime. It would very nice to provide an api to set `checkIfChanged` function as option.

Can it be done in this PR?

**Comment by Mordred** [comment]
@michael-ciniawsky `makeKey` -> `cacheKey`

@Crazymax11 We have different strategy. Instead of directly cloning repo in the docker image (which is slow), we have cloned repository on the build machine. During build we update this repo with `git pull` and copy it into docker context with `COPY`.

File: README.md [hunk 19]
```diff
@@ -47,8 +47,11 @@ module.exports = {
 
 |Name|Type|Default|Description|
 |:--:|:--:|:-----:|:----------|
-|**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored|
-|**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.|
+|**`cacheDirectory`**|`{String}`|`path.resolve('.cache-loader')`|Provide a cache directory where cache items should be stored (used for default read/write implementation)|
+|**`cacheIdentifier`**|`{String}`|`cache-loader:{version} {process.env.NODE_ENV}`|Provide an invalidation identifier which is used to generate the hashes. You can use it for extra dependencies of loaders.  (used for default read/write implementation)|
+|**`cacheKey`**|`{Function(options, request) -> {string}}`|`undefined`|Allows you to override default cache key generator.|
```
Comment (michael-ciniawsky) [inline README.md]: Please finally move this to be the first item in the options table (before `cacheDirectory`) && `-> {string} ` => `-> {String}`

File: src/index.js [hunk 20]
```diff
@@ -35,62 +45,32 @@ function loader(...args) {
...
 function pitch(remainingRequest, prevRequest, dataInput) {
-  const loaderOptions = loaderUtils.getOptions(this) || {};
-  const defaultOptions = {
-    cacheDirectory: defaultCacheDirectory,
-    cacheIdentifier: `cache-loader:${pkgVersion} ${ENV}`,
-  };
-  const options = Object.assign({}, defaultOptions, loaderOptions);
-  const { cacheIdentifier, cacheDirectory } = options;
+  const options = Object.assign({}, defaults, loaderUtils.getOptions(this));
   const data = dataInput;
```
Comment (michael-ciniawsky) [inline src/index.js]:
```diff
- const data = dataInput
```

File: src/index.js [hunk 21]
```diff
@@ -123,4 +103,48 @@ function digest(str) {
...
+    // for performance skip creating directory
+    fs.writeFile(key, content, 'utf-8', callback);
+  } else {
+    mkdirp(dirname, (mkdirErr) => {
+      if (mkdirErr) {
+        callback(mkdirErr);
+        return;
+      }
+
+      directories.add(dirname);
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

File: src/index.js [hunk 22]
```diff
@@ -123,4 +103,48 @@ function digest(str) {
...
+      callback(null, data);
+    } catch (e) {
+      callback(e);
+    }
+  });
+}
+
+function cacheKey(options, request) {
+  const { cacheIdentifier, cacheDirectory } = options;
+  const hash = digest(`${cacheIdentifier}\n${request}`);
```
Comment (michael-ciniawsky) [inline src/index.js]: `\n`

File: src/index.js [hunk 23]
```diff
@@ -35,62 +45,32 @@ function loader(...args) {
...
-  };
-  const options = Object.assign({}, defaultOptions, loaderOptions);
-  const { cacheIdentifier, cacheDirectory } = options;
+  const options = Object.assign({}, defaults, loaderUtils.getOptions(this));
   const data = dataInput;
   const callback = this.async();
-  const hash = digest(`${cacheIdentifier}\n${remainingRequest}`);
-  const cacheFile = path.join(cacheDirectory, `${hash}.json`);
+  const { read: readFn, cacheKey: cacheKeyFn } = options;
+
```
Comment (michael-ciniawsky) [inline src/index.js]:
```js

const data = dataInput

data.cacheKey = cacheKeyFn(options, remainingRequest)
data.remainingRequest = remainingRequest

```

**Comment by Mordred** [comment]
@michael-ciniawsky Changed. 

Can I ask you why are these `\n` changes so important? Current `index.js` has no newlines in function definitions and I tried to match that style :)

**Comment by michael-ciniawsky** [comment]
> Can I ask you why are these \n changes so important? 

Of course, Code Readability. I know it is annoying and doesn't add anything, but that code needs to be maintained and whitespace in particular is one of the cheapest ways to improve readability.

>  Current index.js has no newlines in function definitions and I tried to match that style :)

Yep but it definitely should have and if no one is pesky about it the tendency is to become one big 'BLOB' again. JSDoc is also missing 😛

**Comment by michael-ciniawsky** [comment]
Anything left TODO left here ? @sokra  @webpack-contrib/org-maintainers please review when time :)

**Comment by d3viant0ne** [comment]
Looks like @Mordred has completed all the review comments. I don't personally see any other issues that stand out.

Unless Tobias has an objection, I'd say let it rip. :)

**Comment by hulkish** [comment]
LGTM. Although... I know this applies more to this project as a whole - would be nice to see some tests. 

Not really practical to hold this up for that reason, though.

**Comment by d3viant0ne** [comment]
@hulkish - @michael-ciniawsky setup a base test configuration & we are working through the repos that don't already have suites. Just haven't gotten here yet.

**Pull request merged**